libqpdf/QPDFWriter.cc

```diff
@@ -1274,8 +1274,14 @@ QPDFWriter::willFilterStream(
         QPDFObjectHandle filter_obj = stream_dict.getKey("/Filter");
         if (!m->recompress_flate && !stream.isDataModified() && filter_obj.isName() &&
             (filter_obj.getName() == "/FlateDecode" || filter_obj.getName() == "/Fl")) {
-            QTC::TC("qpdf", "QPDFWriter not recompressing /FlateDecode");
-            filter = false;
+            bool empty_stream = stream_dict.getKey("/Length").getIntValue() == 0;
```

Comment:
This if statement deals with the the specific case of a single flate filter. You want to set filter for empty streams outside the if statements.
 
Also, you should set compress_stream to false to stop the empty output stream from getting recompressed rather than  trying to suppress the writing of `/Filter /FlateDecode` for empty streams.